Added Native Documents page (#137)
ShipEngine/connect-website: native-rendering/documents-designer/data-source.md

---

# Document data source

During template design creation, document data can be accessed using JSONPath syntax. You can use both, the dot notation:

```code
$.shipment.tracking_number
```
or the bracket notation:

```code
$['shipment']['tracking_number']
```
## Notes:
1. For now limited JSONPath syntax can be used. Notice that filter **expressions are not supported** yet.
2. The leading `$.` or `.` or `@` is not optional. One of them have to be be used to differentiate variables and JSONPath syntax.
### Exception:
Special leading char `@` can be use as reference to current object but not in all cases. It depends of document type: e.g. for Package
document type we can use `@package` reference to handle date from each package consecutively from multi packages shipment.

3. Root element can be start from `$.` or `.`
4. Relative nested object path have to be start from `.`
If nested object is started from `$.` it means root element reference. In this case no relation will be consider.
5. **Double dots ..property are NOT handled at all**
6. Wildcard selecting all elements `[*]` can be used in an array only. Regardless of their indexes it means all items of the array.

> 7. **Wildcard selecting all elements `.*` for object (means all properties) is NOT allowed. E.g. `shipment.*`**

Reviewer: ```suggestion
7. **Wildcard `*` cannot be used to select all properties of an object, e.g. `shipment.*`**
```

```suggestion
7. **Is is not possible to select all properties of an object, like `shipment.*`**
```

## JSONPath usage examples
Below JSONPath examples are created for data source object structure:
```code
{
	"shipment": {
		"tracking_number": "123PL",
		"packages": [
			{
				"weight": "12"
			},
			{
				"weight": "02"
			}
		]
	},
	contract {
	    "parameters":{
			"Account.Number": "1234",
			"IsBrexit" : "true"
		}
	}
}
```

## Examples:

### Properties

```code
$.shipment.tracking_number
```

## Arrays

> Can be reference with index or asterisk for each elements iteration

Reviewer: ```suggestion
A specific element can be referenced with an index. `*` wildcard can be used to iterate over a whole collection.
```

Reviewer: Not sure if that suggestion is technically correct or not misleading

```code
$.shipment.packages[0].weight
Sum($.shipment.packages[*].weight)
$['shipment']['packages'][0]['weight']
```

## Dictionaries

> Generally dot and bracket notation can be used but in case when dictionary key contains dot in key name you have to used bracket notation only.

Reviewer: ```suggestion
Generally, a dot and bracket notation can be used, but in case a dictionary key contains a dot, you have to use bracket notation - see the third example.
```

```code
$.contract.parameters.isBrexit
$.contract.parameters['isBrexit']
$.contract.parameters['Account.Number']
```

## Table data

> When you need to design template with dynamic array of data rows (e.g. `$.shipment.packages`) it is necessary to use table element.  

Reviewer: ```suggestion
When you need to design a template with dynamic array of data rows, e.g. `$.shipment.packages`, it is necessary to use table element.  
```

In this case data source has to be defined.  

![Table data](./images/table-data-source.png)<p>
When you work with table you can specify:

### Element by relative path

In this case relative path does not start with `$` but only with dot.

### Elements range [:n], [1,3,5], [0:2], [-n:], [n:]

> Json path elements range is supported as well. E.g. you can select the first n elements of the array or return only selected items if they exist in the array.

Reviewer: ```suggestion
Json path elements range is supported as well. For instance, you can select the first n elements of the array, or return only selected items, if they exist.
```
